cdap-master/src/main/java/io/cdap/cdap/master/upgrade/UpgradeJobMain.java

```diff
@@ -0,0 +1,120 @@
+ * the License.
+ */
+package io.cdap.cdap.master.upgrade;
+
+import com.google.common.base.Throwables;
+import io.cdap.cdap.api.retry.RetryableException;
+import io.cdap.cdap.client.ApplicationClient;
+import io.cdap.cdap.client.NamespaceClient;
+import io.cdap.cdap.client.ProgramClient;
+import io.cdap.cdap.client.ScheduleClient;
+import io.cdap.cdap.client.config.ClientConfig;
+import io.cdap.cdap.client.config.ConnectionConfig;
+import io.cdap.cdap.common.NotFoundException;
+import io.cdap.cdap.common.service.Retries;
+import io.cdap.cdap.common.service.RetryStrategies;
+import io.cdap.cdap.common.service.RetryStrategy;
+import io.cdap.cdap.proto.ApplicationRecord;
+import io.cdap.cdap.proto.NamespaceMeta;
+import io.cdap.cdap.proto.ProgramType;
+import io.cdap.cdap.proto.id.ApplicationId;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.proto.id.ScheduleId;
+import io.cdap.cdap.proto.id.WorkflowId;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * Main class which runs before operator updates CDAP image.
+ * Stops all pipelines and disables all schedules.
+ */
+public class UpgradeJobMain {
+
+  private static final int DEFAULT_READ_TIMEOUT_MILLIS = 90 * 1000;
+  private static final int PORT = 11015;
+  private static final String SCHEDULED = "SCHEDULED";
+
+  public static void main(String[] args) {
+    if (args.length != 1) {
+      throw new RuntimeException(
+        String.format("Invalid number of arguments to UpgradeJobMain. Needed 1, found %d", args.length));
+    }
+    ConnectionConfig connectionConfig = ConnectionConfig.builder()
+      .setHostname("cdap-" + args[0] + "-router")
+      .setPort(PORT)
+      .setSSLEnabled(false)
+      .build();
+    ClientConfig clientConfig =
+      ClientConfig.builder()
+        .setDefaultReadTimeout(DEFAULT_READ_TIMEOUT_MILLIS)
+        .setConnectionConfig(connectionConfig)
+        .build();
+    ApplicationClient applicationClient = new ApplicationClient(clientConfig);
+    ScheduleClient scheduleClient = new ScheduleClient(clientConfig);
+    ProgramClient programClient = new ProgramClient(clientConfig);
+    NamespaceClient namespaceClient = new NamespaceClient(clientConfig);
+    RetryStrategy retryStrategy =
+      RetryStrategies.timeLimit(30, TimeUnit.SECONDS,
+                                RetryStrategies.exponentialDelay(10, 500, TimeUnit.MILLISECONDS));
+
+    try {
+      Retries.supplyWithRetries(
+        () -> {
+          try {
+            for (NamespaceMeta namespaceMeta : namespaceClient.list()) {
+              NamespaceId namespaceId = namespaceMeta.getNamespaceId();
+              for (ApplicationRecord record : applicationClient.list(namespaceId)) {
+                ApplicationId applicationId =
+                  new ApplicationId(namespaceId.getNamespace(), record.getName(), record.getAppVersion());
+                List<WorkflowId> workflowIds =
```

Comment:
there's a lot of nesting here, I think it would be more readable if the logic to suspend all schedules for an app is broken out into a separate method.